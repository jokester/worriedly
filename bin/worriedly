#!/usr/bin/env node

// a dummy entrypoint to run real built js

// eslint-disable-next-line @typescript-eslint/no-var-requires
require('../dist/bin/worriedly').cliMain();
