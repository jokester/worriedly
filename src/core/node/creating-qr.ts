import qrcode from 'qrcode-generator';

/**
 * encode a node.js Buffer into JS UCS-2 string
 * The char codes are used by qrcode like {@code stringEncodedBytes.charCodeAt(charIndex) & 0xff}
 * ([0, 255] range of byte is a subset of legal UCS-2 char code)
 */
export function encodeBuffer(bytes: Buffer): string {
  return String.fromCharCode(...bytes.toJSON().data);
}
