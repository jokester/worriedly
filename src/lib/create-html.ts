export function createHtml(props: { gifDataUri: string }) {
  return ``;
}
