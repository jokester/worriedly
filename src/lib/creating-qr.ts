import qrcode from 'qrcode-generator';

export function createQRFromBytes(bytes: Buffer, errorCorrectionLevel: 'L' | 'M' | 'Q' | 'H' = 'H') {
  const qr = qrcode(0, errorCorrectionLevel);

  /**
   * encode bytes into JS UCS-2 string
   * The char codes are used by qrcode like {@code stringEncodedBytes.charCodeAt(charIndex) & 0xff}
   * ([0, 255] range of byte is a subset of legal UCS-2 char code)
   */
  const stringEncodedBytes = String.fromCharCode(...bytes.toJSON().data);

  qr.addData(stringEncodedBytes, 'Byte');
  qr.make();

  // TODO: what are better size for qr code (considering resolution for print/scan)
  const moduleCount = qr.getModuleCount();

  const gifDataUri = qr.createDataURL(1, 0);

  return {
    moduleCount,
    gifDataUri,
  };
}
