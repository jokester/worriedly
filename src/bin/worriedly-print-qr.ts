import { ArgumentParser } from 'argparse';
import * as fsp from '../ts-commonutil/node/fsp';
import { readStream } from '../ts-commonutil/node';
import { getWinstonLogger } from '../ts-commonutil/logging/winston-logger';

const logger = getWinstonLogger(__filename, 'debug');

export async function worriedlyPrintQr(argv: string[]) {
  const options = parseOptions(argv);
  console.log(options);
  await worriedlyPrintQrMain(options);
}

export interface PrintQrOptions {
  inputFile?: string;
  outputFile?: string;
  verbose?: boolean;

  outputFormat?: 'html' | 'gif';
}

export function parseOptions(argv: string[]): PrintQrOptions {
  const parser = new ArgumentParser({
    version: '0.0.0',
    addHelp: true,
    description: 'TODO: print-qr',
  });

  parser.addArgument(['-i', '--input-file'], {
    dest: 'inputFile',
    metavar: 'INPUT_FILENAME',
    help: 'Input file to read bytes from. Will read from STDIN if unspecified.',
  });

  parser.addArgument(['-o', '--output-file'], {
    dest: 'outputFile',
    metavar: 'FILENAME',
    help: 'Output file to write to. Will write to STDOUT if unspecified.',
  });

  parser.addArgument(['-f', '--output-format'], {
    dest: 'outputFile',
    metavar: 'FORMAT',
    help: '',
  });

  parser.addArgument(['--verbose'], {
    defaultValue: false,
    action: 'storeTrue',
    dest: 'verbose',
    help: `Enable verbose log. Most for development use.`,
  });

  return parser.parseArgs(argv) as PrintQrOptions;
}

async function worriedlyPrintQrMain(options: PrintQrOptions) {
  const inputBytes = options.inputFile ? await fsp.readFile(options.inputFile, { encoding: 'buffer' }) : await readStream(process.stdin);
  if (!options.verbose) logger.silent = true;
  logger.error('got %d bytes', inputBytes.length);
  logger.error('got %o', inputBytes);
}
