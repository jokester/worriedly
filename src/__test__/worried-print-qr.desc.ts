import { inferOutputFormat, SupportedOutputFormat } from '../bin/worriedly-print-qr';
import { createQRFromBytes } from '../lib/creating-qr';
describe('worried-print-qr', () => {
  it('parses option', () => {});

  it('infers output format', () => {
    expect(inferOutputFormat('', '')).toEqual(SupportedOutputFormat.html);

    expect(inferOutputFormat('gIf', 'a.html')).toEqual(SupportedOutputFormat.gif);
    expect(inferOutputFormat('html', undefined)).toEqual(SupportedOutputFormat.html);
    expect(() => inferOutputFormat('v.png', undefined)).toThrowError(/^cannot recognize/i);

    expect(inferOutputFormat('', 'a.htm')).toEqual(SupportedOutputFormat.html);
    expect(inferOutputFormat(undefined, 'b.Gif')).toEqual(SupportedOutputFormat.gif);
    expect(() => inferOutputFormat(undefined, 'x.svg')).toThrowError(/^cannot infer/i);
  });
});

describe('bytes', () => {
  it('we can encode a byte array to UCS-2 string and read it back', async () => {
    const bytes = new Array(256).fill(-1).map((_, index) => index);

    const stringEncoded = String.fromCharCode(...bytes);
    expect(stringEncoded).toHaveLength(256);

    for (let i = 0; i < 256; i++) {
      const readBack = stringEncoded.charCodeAt(i) & 0xff; // same as in qrcode-generator
      expect(readBack).toEqual(bytes[i]);
    }

    const buf = Buffer.from(bytes);
    expect(buf.length).toEqual(bytes.length);
    expect(buf.readUInt8(0)).toEqual(0x000);
    expect(buf.readUInt8(255)).toEqual(0xff);
    const { moduleCount, gifDataUri } = await createQRFromBytes(buf);

    expect(moduleCount).toMatchSnapshot('moduleCount [0x01 .. 0xff]');
    expect(gifDataUri).toMatchSnapshot('gifDataUri of [0x01 .. 0xff]');
  });
});
