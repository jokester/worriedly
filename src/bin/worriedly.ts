import { worriedlyPrintQr } from './worriedly-print-qr';

export async function cliMain() {
  const args = process.argv.slice();

  const [nodeBin, worriedBin, subCommand, ...rest] = args;

  console.debug('worriedly started', args);

  switch (subCommand) {
    case 'print-qr':
      return await worriedlyPrintQr(rest);
    case undefined:
      throw 'no subcommand specified';
  }
}

/**
 * to use with 'npm start:ts' script
 */
if (require.main === module) {
  setTimeout(cliMain);
}
