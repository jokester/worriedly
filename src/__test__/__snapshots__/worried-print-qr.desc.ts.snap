// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bytes we can encode a byte array to UCS-2 string and read it back: gifDataUri of [0x00 .. 0xff] 1`] = `"data:image/gif;base64,R0lGODdhVQBVAIAAAAAAAP///ywAAAAAVQBVAAAC/4QPEcm5epSE0MhoV3y6ZadRHgKK5QZmk4V6V/hgjsvAZYh1KUdNPgmszX4sUiX2knWQiMoIqUvuesfUTVaLFZdZ1svnzPF0jfBnOlTRrkvcJ7qbNd9DJ/W3upNzQqa/D/gnGEioZQS2IseRBQNVFFQnaPeGmEf25XZ1uOjFBekiknmpgkmkmEf6ZHiziPjImVaGJRtLZ3IkNdnZ1orKFwlb9hXUo0S3iwe2NVYqSmPDhZOo5stIbbrZ+BvVm8y6JctM/EjstnqSxMPWOme2VzVda4ZcvKzsVTuSZoi54aiFy585gAHtXGgSxk+ogy2MhSKXECK0iLlsXFMmDlychf+pBHpr51HarVLPchWqQw4ZPm7CrFCEU+hfrHgZGd3qqCvcGXXMpKGzNg3KKm38FOHEF8mYvo0mjgHsk0kjimIMMRo5BErdCYRqlHjqBs0oroN3XK7pwrMrLSxp1Wr11XanU3Bv8bCiSjaYOVI8JxUk+DNvVX9sxEA95SqfSJbCEhGNi/RqWIT/Ao6z6ZjhUZe6Bl+8C0erHJWLUw6qchfpxH0gl47mJJMyZrTJGtUbbe0U4UuW52HldWZObp2SKfHe8zoO7XTNdK+0nbUs7ldBVbU4p3kUOm3AAm+b1TLndX5Udn0apefpdKB8VZm0GPoaeyDSbaGqSJIWfM2yWXP/dUeXYz+twYtI1PmUDlstLSgKNwqFUxsfRWE3BmFWzEQRhMOUsxSCy6ykH1BENfThLHiRxFw9qAkFCkQ2RQYiSLdhZJV5u7Gj1oXstOWXgy26RwRg2jkiVV1pQYZYjHjZJ86IwRGZ2VGw+CRXT3rB9oRpkqnIVYmN/dbTgoVBdVxVQmUnH2dUmaVlUjTRN9aNg4XWD1br1DdQNPaQqeeKlYy3CYlNoSZGQamUR1ZKQhiVIzwC2taPJYvdRl02XQrHhJrHHapnNoNSKdx2CFLp6H7RmKnekMBtuChKhoJ6D6MgHpnoaaui5N1b33Ak35yMpjiPK+4Ea0l4plGq2jsE/0U51WuwJsihgO3pFik8WyVJaq3mPSYFU1kqOVWFVbYo3mGewadcsYuCitaLMzkXJzXN1qSJmL9u+Cga1vIWqqbSVmqPtetu+6AtlJCJKUxmOstpq4BJ2S1MrtEV6pse6jPUwe9mWlp+TTFsn1RuGqRQrSwtC9x/jU0pG5aEIpcvYlFxRuBD3KG7pGitedTwd87WSR6EyWUUMYY1n7iZWDC/ImvEUWWs9MqxYfyfuGzeZ1C99XpiYVbB4fZXeWgYvBO+Cz3UcoXqMsyffmMhOsw5y4nl1TcOi8o1i71cxbdgfmMNJ6ofEwLangbGuSbA4dnMX03x+feOdTOPM2E8hWXSqJTQgrlLZ8oztrwewl10AqNoiZbuzFNEjlnS6LuJjNwUVnfX9qn6qujbvpNDIq7Y8xpmIY6uR3eojltBifwzcgZZOusR9su1iTwLPNLyCfVIKudMWmQn3Tuit+6g7b29JpoFAAA7"`;

exports[`bytes we can encode a byte array to UCS-2 string and read it back: moduleCount [0x00 .. 0xff] 1`] = `85`;
